pkg/icingadb/sla.go

```diff
@@ -0,0 +1,80 @@
+package icingadb
+
+import (
+	"context"
+	"crypto/rand"
+	"github.com/icinga/icingadb/pkg/com"
+	"github.com/icinga/icingadb/pkg/contracts"
+	v1 "github.com/icinga/icingadb/pkg/icingadb/v1"
+	"github.com/icinga/icingadb/pkg/types"
+	"github.com/icinga/icingadb/pkg/utils"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+	"time"
+)
+
+type SlaHistoryTrail struct {
+	v1.EntityWithoutChecksum `json:",inline"`
+	EnvironmentId            types.Binary    `json:"environment_id"`
+	ObjectType               string          `json:"object_type"`
+	HostId                   types.Binary    `json:"host_id"`
+	ServiceId                types.Binary    `json:"service_id"`
+	EventTime                types.UnixMilli `json:"event_time"`
+	EventType                string          `json:"event_type"`
+}
+
+func CheckableToSlaTrailEntities(ctx context.Context, checkables <-chan contracts.Entity, eventType string) (<-chan contracts.Entity, <-chan error) {
+	entities := make(chan contracts.Entity)
+	g, ctx := errgroup.WithContext(ctx)
```

Comment:
Admittedly this is an established pattern. To create new groups inside functions. But I like re-using existing ones -as in ErrgroupReceive- better.